spec/12-the-scala-standard-library.md

```diff
@@ -642,6 +642,12 @@ object Predef {
   def classOf[T]: Class[T] = null
    // this is a dummy, classOf is handled by compiler.
 
+  // valueOf -----------------------------------------------------------
+
+  /** Retrieve the single value of a type with a unique inhabitant. */
+  @inline def valueOf[T](implicit vt: ValueOf[T]): T {} = vt.value
```

Comment:
> * Adding ClassTag[T] to classOf would be wrong, as classOf requires a ClassTag only in some cases, it doesn't require a ClassTag when the argument is a constant type.
> * Same for Array[T].

And in both those cases, an implicit `ClassTag[T]` can be materialized when `T` is a constant type. So demanding an `implicit ClassTag[T]` is just as correct.

> Nothing in that requires adding ClassTag to classOf

Sure, but then it requires `classOf` to be magic. You either need `classOf` to take an implicit `ClassTag` or to be magic. I prefer not-magic. Now, as per @nafg's comment, `classOf` needs to be magic for an unrelated reason. But `valueOf` doesn't need to.

> and it would be wrong for the reasons mentioned.

Nope, see my first answer.

> This eliminates the argument for

It doesn't. Less magic is still better.

> as the only question remaining now is whether we want to have valueOf consistent with classOf or doing its own thing. I prefer consistency.

There is the second question: more magic or less magic.

> That's why the ValueOf constructor needed to be private (see comment below).

I agree that the constructor of `ValueOf` should be private. This is indeed necessary to guarantee that an existing instance of `ValueOf[T]` is correct by construction. When that is guaranteed, there cannot be any harm done (whether involuntarily or maliciously) by passing an explicit `ValueOf[T]` to `valueOf[T]`. There is therefore no need for the compiler to take protection measures for `valueOf[T]`, so it need not be magic, so it can be defined in user-space with a perfectly normal `implicit ValueOf[T]`.